### ISSUE TITLE
[Bug]: Taz Express - very-broken using Project64 video-plugin; somewhat broken with Jabo Direct3D8

### ISSUE THREAD

**Issue by dftf-stu** [project64/project64#2256]
### Expected behavior?

1. The game should not flicker in places
2. All "story text" should be visible
3. The "Pause menu" (press START during gameplay) should be visible

### Actual behavior?

When using the **Project64** video plugin:
- Title screen: after you press START, the screen flickers
- "Select game" screen: after you select a save-slot, the screen flickers
- After the "select game" screen, there should be some story-text before the brief cut-scene, but this does not appear
- On the "Taz Out Back" loading-screen (with the orange background) there should be some story-text, but this does not appear
- "Pause menu": does not display; the screen just flickers with a black-box in the centre (press START during gameplay)
- **NOTE:** enabling the "interpreter core" option makes no-difference

When using the **Jabo's Direct3D8** video plugin:
- During all opening screens (such as "Select Language") the background is mostly-black, when it should be orange
- Title screen: obvious texture-rendering issues are visible during the demo
- All of the story-text is now visible
- The "Pause menu" is now visible
- **NOTE:** enabling the "interpreter core" option makes no-difference

When using the **GLideN64** video plugin:
- Most issues not present, though occasional minor graphical-glitches are still present

### Steps to reproduce the problem

Simply run the game up-to the start of the first level using each of the video plugins and observe how-broken the "Project64" default one is compared with the other two -- you don't see any of the introduction "story-text", and the "Pause Menu" is broken.

### Error messages?

N/A

Here is a screenshot:

![TazExpressIssues](https://user-images.githubusercontent.com/83965343/180194233-cf06823a-7477-40e3-8a14-4cc91d001340.png)

### What game were you playing?

Title: Taz Express
Game type: ROM only (non Disk-Drive)
ROM information: MD5=E53B8573D54E5B4694A77FD1DD3FFFB0 | CRC1=0xAEBCDD54 | CRC2=0x15FF834A | CIC=CIC-NUS-6102 
Save type: Default
Original or modified ROM?: Original

### Specifications and build details

Project64: 4.0.0.5846
OS: Windows 10 Home (64-bit) Version 21H2 Build 19044.1826
CPU: Intel Core i7-5600U
RAM: 16GB
Graphics: Intel HD 5500 + AMD Radeon R7 M260X

### Using included plugins

- [X] I can confirm I am using the included RSP, audio, input, and graphics plugins

### Plugins used while the issue occurred

Graphics plugin (and HLE or LLE): tested all three, only on HLE mode
Audio plugin: default
Input plugin: default
RSP plugin: default

### Copyright Compliance

- [X] There are no infringing files attached to this bug report.

**Comment by Squall-Leonhart** [comment]
Jabo is discontinued / dead.

Pj64Video, disable framebuffer emulation

**Comment by dftf-stu** [comment]
@Squall-Leonhart Thanks for the reply.

> "Jabo is discontinued / dead."

If it's no-longer supported, maybe the developers should consider not shipping it with the emulator by-default then, or at-least add some sort of warning when that plug-in is selected to advise that?  I'm honestly staggered by the amount of things I've been told here since logging bugs which are not mentioned anywhere... yet people are just expected to somehow know?

> "Pj64Video, disable framebuffer emulation"

Assuming you mean Options > Graphics Settings > Emulation Settings tab > un-tick "Enable frame buffer emulation" then this partially resolves the issue -- the text is now visible -- but the screen still constantly flickers.

Are these settings correct?

![TAZ_STNS](https://user-images.githubusercontent.com/83965343/180232304-f9a561b5-ac2b-4cc9-ab64-fc9bd574925b.png)

**Comment by flagrama** [comment]
> If it's no-longer supported, maybe the developers should consider not shipping it with the emulator by-default then, or at-least add some sort of warning when that plug-in is selected to advise that?

It's literally the only plugin of the included ones that works for people that don't have OpenGL support. That's the only reason it's included still. Zilmar is always trying to keep compatible options in PJ64.

>  I'm honestly staggered by the amount of things I've been told here since logging bugs which are not mentioned anywhere... yet people are just expected to somehow know?

Everybody who has used PJ64 since at least 1.6 knows Jabo's plugin is buggy, but [Jabo's plugin is closed source so nobody can work on it since Jabo stopped](http://www.jabosoft.com/articles/114) and of course speedrunners would rather blacklist an emulator for the slightest mistake rather than actually make them better so even to this day 17 years later tell people to use 1.6 where that's the only choice by default. So this information gets repeated quite a lot.

PJ64 has been around for 20 years. People are more used to answering the question as it comes up rather than writing it down where nobody reads it anyway and it needs to be answered when a question comes up anyway. Especially since many of the places where stuff like this _was_ written down doesn't exist anymore, or is further fragmented due to people thinking ["let's do thing, but better"](https://xkcd.com/927/) (reddit, wikis, forums, and other types of websites)

**Comment by Squall-Leonhart** [comment]
So this is an issue in PJ64Video, and will likely require reducing hardware compatibility further to move on up to more recent GlideN64 code.

**Comment by DerekTurtleRoe** [comment]
Yeah, much of what I know about N64 emulation has come from being in the scene for nearly 2 decades. I'm trying to improve the situation as much as I can with better documentation, better comments/notes, and improving the actively maintained plugins and emulators in any way I can. But there is only so much I can do all by myself, unfortunately.

All closed-source N64 emulators and plugins are either no longer maintained, or if they are then only the main developer can maintain them, so in my opinion I don't consider them when doing my own testing because if the developer abandons the project, etc etc etc then there isn't anything I can do as a user. So that's why GLideN64 is distributed with the emulator along with Project64 video and all of Zilmar's plugins. Those are all maintained still and we have people working on them, or at the very least we can look at the code and see what is happening.

**Comment by dftf-stu** [comment]
@flagrama: old versions of Project64 (back in the early 2000s) used to have a "Notes" column in the ROM Browser by-default, which used to explain any setting changes, or specific plug-ins, a game needed.  By-default, this is no-longer displayed and any ROM I've tested all appear in green text saying "Compatible".  For all such games, it is not unreasonable to expect that should mean "it works out-of-the-box", not "I need to go on the Discord before I play and ask people what settings must be tweaked within the global-settings and the ROM itself to make it work"?  I never have to do this for any-other emulator I use...

@DerekTurtleRoe: I would assume that "Project64 video" is owned by the current developers of Project64, given the name?  And as such I'd expect that would be the most-stable option to go-with as it should be the most-tested against, no?

**Comment by dftf-stu** [comment]
So on the compatibility-front, if certain settings need to be changed for certain ROMs, and people here know exactly which ones need changing... then why not just do that automatically?

**DuckStation** has a file called "compatibility.xml", which sometimes includes a note that will appear alongside a game:
> [entry code="SLUS-01270" compatibility="5" region="NTSC-U" title="Medal of Honor - Underground (USA)"]
> [compatibility]No Issues[/compatibility]
> [version-tested]0.1-2499-g770dd5a6[/version-tested]
> [comments]Last sound of the BIOS sound keeps looping until the first loading screen in the game starts. It's related with CD-ROM speeds and can be avoided with "Fast Boot" option on (Issue #407).[/comments]

And it also includes a file called "gamesettings.ini" that apply hacks required for specific titles, with entries such as:
> #SLUS-01222 (Colin McRae Rally 2.0 (USA) (En,Fr,Es))
>[SLUS-01222]
>ForceRecompilerLUTFastmem = True
>DisplayActiveStartOffset = 64
>DisplayActiveEndOffset = 68

**PCSX2** also has a similar file called "GameIndex.yaml", which also provides hacks necessary to make some games work:
>CPCS-01005:
>name: "Gun Survivor 4 - Biohazard - Heroes Never Die [with GunCon2]"
>region: "NTSC-J"
>gsHWFixes:
>halfPixelOffset: 1 # Fixes character offset with flashlight and blurriness.
>alignSprite: 1 # Fixes vertical lines.
>roundSprite: 2 # Fixes font artifacts.
>mergeSprite: 1 # Fixes flame-like bleeding.

Can't something similar be done for N64 games to apply known settings changes that are required, rather than leave users having to ask on here or Discord on how to make things work?

**Comment by Squall-Leonhart** [comment]
"Owned" is excessive wording, its a snapshot of Glide64 and has all the inherent issues that Glide64 at the time had, 

the next evolution for the RDP would be either making GlideN64 the default, or forking it.

**Comment by project64** [comment]
at one stage smiff use to maintain the compatibility list and fix up notes, he is sorely missed in the project. I do not have the time to test games and update that, so the notes are turned off by default cause more often then not they are wrong.

**Comment by dftf-stu** [comment]
@Squall-Leonhart: "the next evolution for the RDP would be either making GlideN64 the default, or forking it"

Personally I'd say just make a software-renderer, like many other emulators from the Saturn/PS1/N64 and PS2/Gamecube/Xbox era have, and be done.  Then you can avoid all issues caused by poor OpenGL implementations, and bad driver releases.

I'm able to get full 100% speed on PCSX2 using the "Software Renderer" mode on my newer laptop -- if PS2 games can be rendered fully only on the CPU, there's no-reason I can see that N64 games couldn't.

**Comment by flagrama** [comment]
A software renderer is something I believe zilmar plans to make as part of his accuracy improvements, but only for testing because it is highly unlikely it will be at all performant if it's going to be accurate. The fact that even [the hardware accelerated accurate plugins](https://github.com/ata4/angrylion-rdp-plus) can be quite slow leaves not much hope for performance from a software renderer.

I don't know anything about how PCSX2 has been developed, but I cannot believe the software renderer can be anywhere close to accurate if it's fast, a CPU and a GPU have fundamentally different architecture. It's not like the 2D consoles where the way to display video was primarily just another regular processor. Maybe if the number of parallel operations was low enough on it that newer high-core-count processors have exceeded that number.

N64 of course is complicated in its own way because it didn't go with a completely normal GPU architecture. It went with a MIPS co-processor that itself is quite complex. The 2 major parts of it are the Reality Signal Processor (RSP) (which itself has 2 co-processors) and the Reality Display Processor (RDP) as you may be familiar with messing with the plugins.

If you want some technical details, there's stuff scattered all over the place, but [here is one example](https://www.copetti.org/writings/consoles/nintendo-64/#graphics)

**Comment by DerekTurtleRoe** [comment]
@dftf-stu Well, first of all "just make a software renderer" is kind of dismissive of the fact that making any renderer is difficult, let alone an entire renderer in software. Coupled with the fact that the N64 is one of the most difficult consoles to emulate currently, that isn't really a good solution in the long run, or in the short term either. A software renderer will be really nice for debugging and such, and I will be excited for it, but currently we don't even have reference-quality accurate HLE or LLE RDP emulation, and we need to do a lot more research before it's plausible I think.

One thing I will say (no offense to the PCSX2 team, they've done amazing work and I have been privileged enough to work with them on fixing things on occasion) is that the software renderer, while pretty accurate, isn't 100% accurate. Not only that but the hardware renderers also need a lot of work as well. Specifically I think the instruction cache and some other random bugs cause certain games to have texture issues or horrid performance. They have made amazing strides in the last 6 - 12 months! They've basically rewritten huge parts of the emulator, switched from WxWidgets to Qt, rebuilt the Mac port, etc etc etc. But you also have to take into consideration the amount of contributors they have and how many people bought PS2s and loved them, either now or when they were kids. Sure, people owned the N64 as a kid, but due to many factors, including that the N64 is difficult to research, a long history of incorrect assumptions about how the N64 works, and a lack of documentation and research on the console, along with many other things, mean that overall less people are interested in N64 research and development then the PS2, at least for the most part. Play! is trying to be more accurate and still is very early in development, and it has performance issues as well.

To answer your question about Glide64/Project64 video, the most "stable" plugin is probably either GLideN64 or Parallel-RDP. Those are also probably the most accurate as well. We include GLideN64 because it is one of the best all-around RDP plugins. Features, compatibility, stability, etc.

Another note is that we are trying to make the emulator more accurate, and in doing so inevitably the emulator will sometimes break and games won't work. Just running many of the available test ROMs for N64 on Project64 causes a lot of failures, skipped tests, and crashes/hangs. The CPU emulation, memory, and RSP are the most common things that fail (at least in my testing). Not that the RDP emulation is perfect, but overall it doesn't fail as often. I'm confident that when the CPU and memory are more accurate a lot of games will have far fewer issues. RSP emulation will also help with that as well I believe.

In regards to the ROM browser and the notes column, IMO the notes column should be removed eventually since like Zilmar said it will be difficult to maintain and eventually (hopefully) it won't be needed. The goal is to not need the notes and status columns of the ROM browser, because it is accurate enough to assume that all games will work.

One of the struggles of making an emulator with plugin support is that the core is maintained by the team, and the plugins are often developed by people who may not even know the Project64 architecture or code base, and they can even theoretically do it in programming languages that they choose, so sometimes you can't just look at code for one thing and know what is going on if you don't know how to read multiple programming languages.

Anyway, TL;DR we are in a transition period where things are going to break, just FYI. Letting us know about things that are broken always helps since I can't test every game every commit. I try to test things with the most accurate plugins so we know if core issues come up. Hopefully soon I can get my N64 setup and get some work done on Project64, been having a lot of real life stuff to deal with.

**Comment by Jj0YzL5nvJ** [comment]
> The fact that even [the hardware accelerated accurate plugins](https://github.com/ata4/angrylion-rdp-plus) can be quite slow leaves not much hope for performance from a software renderer.

That an application uses multithreading does not make it "hardware accelerated". I'm pretty sure Angrylion-Plus doesn't even use buffer storage (VBO, PBO, etc).

GLideN64 is somewhat "hardware accelerated" (VBO, PBO, FBO, etc.) but `fragment_shader_interlock` throws performance out the window (coverage and depth compare are the worst offenders), Angrylion-RDP-X uses some kind of CUDA implementation so is also hardware accelerated (GPGPU), ParaLLEl RDP uses Vulkan compute shaders (SSBO, UBO) so is hardware accelerated.
The less work the CPU and RAM are required to do, and the better you use the GPU and VRAM, the more hardware acceleration you will be using.

For me, hardware acceleration outside of video acceleration (VAAPI, NVDEC, DXVA2, VDPAU, QSV, CUDA, etc) will always be questionable.

**Comment by flagrama** [comment]
yeah, I left one mistake in the comment because it wasn't worth correcting, already put too much effort in the comment in the first place.
